src/main/java/com/fasterxml/jackson/databind/introspect/AnnotatedFieldCollector.java

```diff
@@ -70,9 +71,12 @@ private Map<String,FieldBuilder> _findFields(TypeResolutionContext tc,
         // Let's add super-class' fields first, then ours.
         fields = _findFields(new TypeResolutionContext.Basic(_typeFactory, parent.getBindings()),
                 parent, fields);
+
+        JsonFormat format = cls.getAnnotation(JsonFormat.class);
+        boolean isHandledAsObject = format != null && format.shape() == JsonFormat.Shape.OBJECT;
```

Comment:
Yeah, "shape-shifting" support is a big mess, since it has to start from different places, in case of per-property override (Class annotations are easier).

I am not sure how or why they break in this case.